### PR TITLE
Refactor memory to improve API & fix failing tests

### PR DESCRIPTION
**Pull request by danielperano**
Misc. cleanup to internal & external API for the Memory class, targeting fixing all failing memory init tests.

### CODE REVIEW

**Comment by andreaTP** [comment]
Thanks for picking this up @danielperano it was on my todo but happy to see you taking over!
Please, while looking at this, I think we should debug the Data segments initialization of the memory which I think is not completely correct ATM.

File: runtime/src/main/java/com/dylibso/chicory/runtime/Machine.java [hunk 1]
```diff
@@ -320,140 +320,140 @@ void eval(List<Instruction> code) throws ChicoryException {
                     case I32_LOAD:
                         {
                             var ptr = (int) (operands[1] + this.stack.pop().asInt());
-                            var val = instance.getMemory().getI32(ptr);
+                            var val = instance.getMemory().readI32(ptr);
```
Comment (bhelx) [inline runtime/src/main/java/com/dylibso/chicory/runtime/Machine.java:323]: nice, i think this is a little clearer

File: runtime/src/main/java/com/dylibso/chicory/runtime/Memory.java [hunk 2]
```diff
@@ -120,228 +116,191 @@ public void reinstantiate() {
...
     }
 
-    public void put(int offset, byte[] data) {
+    public void write(int addr, byte[] data) {
+        write(addr, data, 0, data.length);
+    }
+
+    public void write(int addr, byte[] data, int offset, int size) {
         try {
-            // System.out.println("mem-write@" + offset + " " + data);
```
Comment (bhelx) [inline runtime/src/main/java/com/dylibso/chicory/runtime/Memory.java:183]: Thank you for fixing this! Couldn't find the right API for this.

Comment (danielperano) [reply runtime/src/main/java/com/dylibso/chicory/runtime/Memory.java:183]: Yeah, `ByteBuffer` is tricky with that API inconsistency - I seem to recall running into it before, and it's one of the reasons I don't particularly like using `ByteBuffer` outside of I/O (though I think its use here is as good as we can do outside of using Panama's native memory API, which isn't even finalized until JDK 22).

File: wasm/src/main/java/com/dylibso/chicory/wasm/Parser.java [hunk 3]
```diff
@@ -643,22 +643,24 @@ private static DataSection parseDataSection(
         var dataSegments = new DataSegment[(int) dataSegmentCount];
 
         for (var i = 0; i < dataSegmentCount; i++) {
-            var memidx = readVarUInt32(buffer);
-            // if memory idx == 0, then this is an active datasegment
-            // if memory idx == 1, then this is a passive datasegment
-            // TODO won't this break once we support multi-memory?
-            if (memidx == 0) {
+            var mode = readVarUInt32(buffer);
```
Comment (andreaTP) [inline wasm/src/main/java/com/dylibso/chicory/wasm/Parser.java:646]: Nice one!
This covers all the cases now:
https://webassembly.github.io/spec/core/binary/modules.html#data-section

🎉  do you want to change the explicit throw of the exception to an assertion?

**Review by andreaTP — APPROVED** [review]
A minor nitpick but LGTM it's a good step forward!

We still miss to move the memory initialization after instantiation and correctly apply it to  imported memory, but it can be done in a separate PR 👍

File: wasm/src/main/java/com/dylibso/chicory/wasm/Parser.java (hunk 3 above)
Comment (danielperano) [reply wasm/src/main/java/com/dylibso/chicory/wasm/Parser.java:646]: For something like this I'd prefer the throw - if someone tries to parse a newer version of a binary (or an incorrectly version labeled binary, or a malformed binary) then it's conceivable that parsing might make it this far.

**Comment by danielperano** [comment]
> A minor nitpick but LGTM it's a good step forward!
> 
> We still miss to move the memory initialization after instantiation and correctly apply it to imported memory, but it can be done in a separate PR 👍

I think that's a pretty tiny change - what did you have in mind for that? Just move the initialization call from the module constructor to `instantiate()`?

File: wasm/src/main/java/com/dylibso/chicory/wasm/Parser.java (hunk 3 above)
Comment (andreaTP) [reply wasm/src/main/java/com/dylibso/chicory/wasm/Parser.java:646]: just a personal preference, not fussy on this 🤷

**Review by andreaTP — APPROVED** [review]
LGTM, feel free to merge.

> I think that's a pretty tiny change - what did you have in mind for that? Just move the initialization call from the module constructor to instantiate()?

Yes and making sure that initialization applies also to imported memories to fix #127 
Let's merge this one and build on top 👍

**Pull request merged**